src/shared/clean-ipc.c

```diff
@@ -346,44 +361,91 @@ static int clean_posix_mq(uid_t uid, gid_t gid) {
         return log_warning_errno(errno, "Failed to read /dev/mqueue: %m");
 }
 
-int clean_ipc(uid_t uid, gid_t gid) {
+static int clean_ipc_internal(uid_t uid, gid_t gid, bool clean) {
         int ret = 0, r;
 
+        /* If either of UID/GID are specified as root, then disable cleaning up for it. If we are not supposed to clean
+         * but just search, then say the UID/GID is busy. */
+        if (uid == 0) {
+                if (!clean)
+                        return -EBUSY;
+
+                uid = UID_INVALID;
+        }
+        if (gid == 0) {
+                if (!clean)
+                        return -EBUSY;
+
+                gid = GID_INVALID;
+        }
+
         /* Anything to do? */
         if (!uid_is_valid(uid) && !gid_is_valid(gid))
                 return 0;
 
-        /* Refuse to clean IPC of the root user */
-        if (uid == 0 && gid == 0)
-                return 0;
+        r = clean_sysvipc_shm(uid, gid, clean);
+        if (r < 0) {
+                if (!clean)
+                        return r;
 
-        r = clean_sysvipc_shm(uid, gid);
-        if (r < 0)
                 ret = r;
+        }
+
+        r = clean_sysvipc_sem(uid, gid, clean);
+        if (r < 0) {
+                if (!clean)
+                        return r;
 
-        r = clean_sysvipc_sem(uid, gid);
-        if (r < 0)
                 ret = r;
+        }
+
+        r = clean_sysvipc_msg(uid, gid, clean);
+        if (r < 0) {
+                if (!clean)
+                        return r;
 
-        r = clean_sysvipc_msg(uid, gid);
-        if (r < 0)
                 ret = r;
+        }
+
+        r = clean_posix_shm(uid, gid, clean);
+        if (r < 0) {
+                if (!clean)
+                        return r;
 
-        r = clean_posix_shm(uid, gid);
-        if (r < 0)
                 ret = r;
+        }
+
+        r = clean_posix_mq(uid, gid, clean);
+        if (r < 0) {
+                if (!clean)
+                        return r;
 
-        r = clean_posix_mq(uid, gid);
-        if (r < 0)
                 ret = r;
+        }
 
         return ret;
 }
 
+int clean_ipc(uid_t uid, gid_t gid) {
+        return clean_ipc_internal(uid, gid, true);
+}
+
 int clean_ipc_by_uid(uid_t uid) {
-        return clean_ipc(uid, GID_INVALID);
+        return clean_ipc_internal(uid, GID_INVALID, true);
 }
 
 int clean_ipc_by_gid(gid_t gid) {
-        return clean_ipc(UID_INVALID, gid);
+        return clean_ipc_internal(UID_INVALID, gid, true);
+}
+
+int search_ipc(uid_t uid, gid_t gid) {
+        return clean_ipc_internal(uid, gid, false);
+}
+
+int search_ipc_by_uid(uid_t uid) {
```

Comment:
At least some of those could be defined as inline in the header, e.g. as `return search_ipc(uid, GID_INVALID)`. This would reduce the boilerplate code.

Reply:
well, these we take function ptrs of these functions, in manager_unref_uid() and manager_unref_gid(). Doing inline functions are OK in that case, as gcc can handle that, but we win nothing in that case, as the functions cannot possibly be inlined, hence I don't think we should do that.

I'll drop the other wrapper functions however, we don#t really need them

Reply:
Ah, I didn't think about taking the pointer. You're right of course.